drivers/led/sn32/rgb_matrix_sn32f24xb.c

```diff
@@ -0,0 +1,426 @@
+                cfg->channels[19].mode = PWM_OUTPUT_ACTIVE_LOW;
+                chan_col_order[i] = 19;
+                break;
+
+            case C12:
+                cfg->channels[20].pfpamsk = 1;
+            case B4:
+                cfg->channels[20].mode = PWM_OUTPUT_ACTIVE_LOW;
+                chan_col_order[i] = 20;
+                break;
+
+            case C13:
+                cfg->channels[21].pfpamsk = 1;
+            case B5:
+                cfg->channels[21].mode = PWM_OUTPUT_ACTIVE_LOW;
+                chan_col_order[i] = 21;
+                break;
+
+            case C14:
+                cfg->channels[22].pfpamsk = 1;
+            case B6:
+                cfg->channels[22].mode = PWM_OUTPUT_ACTIVE_LOW;
+                chan_col_order[i] = 22;
+                break;
+
+            case C15:
+                cfg->channels[23].pfpamsk = 1;
+            case B7:
+                cfg->channels[23].mode = PWM_OUTPUT_ACTIVE_LOW;
+                chan_col_order[i] = 23;
+                break;
+        }
+    }
+}
+void rgb_callback(PWMDriver *pwmp);
+
+void shared_matrix_rgb_enable(void) {
+    pwmcfg.callback = rgb_callback;
+    pwmEnablePeriodicNotification(&PWMD1);
+}
+
+void shared_matrix_rgb_disable_pwm(void) {
+    // Disable PWM outputs on column pins
+    for(uint8_t y = 0; y < LED_MATRIX_COLS; y++) {
+        pwmDisableChannel(&PWMD1,chan_col_order[y]);
+    }
+}
+
+void shared_matrix_rgb_disable_leds(void) {
+    // Disable LED outputs on RGB channel pins
+    for (uint8_t x = 0; x < LED_MATRIX_ROWS_HW; x++) {
+        writePinLow(led_row_pins[x]);
+    }
+}
+
+void update_pwm_channels(PWMDriver *pwmp) {
+    matrix_row_t row_shifter = MATRIX_ROW_SHIFTER;
+    for(uint8_t col_idx = 0; col_idx < LED_MATRIX_COLS; col_idx++, row_shifter <<= 1) {
+        #if(DIODE_DIRECTION == ROW2COL)
+            // Scan the key matrix column
+          matrix_read_rows_on_col(current_matrix,col_idx,row_shifter);
+        #endif
+        uint8_t led_index = g_led_config.matrix_co[row_idx][col_idx];
+        // Check if we need to enable RGB output
+        if (led_state[led_index].b != 0) enable_pwm |= true;
+        if (led_state[led_index].g != 0) enable_pwm |= true;
+        if (led_state[led_index].r != 0) enable_pwm |= true;
+        // Update matching RGB channel PWM configuration
+        switch(current_row % LED_MATRIX_ROW_CHANNELS) {
+        case 0:
+                if(enable_pwm) pwmEnableChannelI(pwmp,chan_col_order[col_idx],led_state[led_index].b);
+            break;
+        case 1:
+                if(enable_pwm) pwmEnableChannelI(pwmp,chan_col_order[col_idx],led_state[led_index].g);
+            break;
+        case 2:
+                if(enable_pwm) pwmEnableChannelI(pwmp,chan_col_order[col_idx],led_state[led_index].r);
+            break;
+        default:
+            ;
+        }
+    }
+}
+void rgb_callback(PWMDriver *pwmp) {
+    // Disable the interrupt
+    pwmDisablePeriodicNotification(pwmp);
```

Comment:
Apparently calling `pwmDisablePeriodicNotification()` is wrong here, because the PWM callback is actually called from the interrupt handler, and in that state most ChibiOS APIs are not allowed. You would need to call `chSysLockFromISR()` earlier, and then use `pwmDisablePeriodicNotificationI()`.